language/move-prover/stackless-bytecode-generator/src/livevar_analysis.rs

```diff
@@ -0,0 +1,213 @@
+/// before being overwritten.
+#[derive(Default)]
+pub struct LiveVarAnnotation(BTreeMap<CodeOffset, BTreeSet<TempIndex>>);
+
+pub struct LiveVarAnalysisProcessor();
+
+impl LiveVarAnalysisProcessor {
+    pub fn new() -> Box<Self> {
+        Box::new(LiveVarAnalysisProcessor())
+    }
+}
+
+impl FunctionTargetProcessor for LiveVarAnalysisProcessor {
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        func_env: &FunctionEnv<'_>,
+        mut data: FunctionTargetData,
+    ) -> FunctionTargetData {
+        let offset_to_live_refs = if func_env.is_native() {
+            // Native functions have no byte code.
+            LiveVarAnnotation(BTreeMap::new())
+        } else {
+            let cfg = StacklessControlFlowGraph::new_backward(&data.code);
+            LiveVarAnnotation(LiveVarAnalysis::analyze(&cfg, &data.code))
+        };
+        // Annotate function target with computed life variable data.
+        data.annotations
+            .set::<LiveVarAnnotation>(offset_to_live_refs);
+        data
+    }
+}
+
+struct LiveVarAnalysis();
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct LiveVarState {
+    livevars: BTreeSet<TempIndex>,
+}
+
+impl LiveVarState {
+    fn remove(&mut self, vars: &[TempIndex]) -> bool {
+        let mut removed = false;
+        for v in vars {
+            if self.livevars.remove(v) {
+                removed = true;
+            }
+        }
+        removed
+    }
+
+    fn insert(&mut self, vars: Vec<TempIndex>) {
+        for v in vars {
+            self.livevars.insert(v);
+        }
+    }
+
+    fn reset(&mut self) {
+        self.livevars = BTreeSet::new();
+    }
+}
+
+impl LiveVarAnalysis {
+    fn analyze(
+        cfg: &StacklessControlFlowGraph,
+        instrs: &[Bytecode],
+    ) -> BTreeMap<CodeOffset, BTreeSet<TempIndex>> {
+        let initial_state = LiveVarState {
+            livevars: BTreeSet::new(),
+        };
+        let mut analyzer = LiveVarAnalysis {};
+        let state_map = analyzer.analyze_function(initial_state, &instrs, cfg);
+        analyzer.post_process(cfg, instrs, state_map)
+    }
+
+    fn post_process(
+        &mut self,
+        cfg: &StacklessControlFlowGraph,
+        instrs: &[Bytecode],
+        state_map: StateMap<LiveVarState>,
+    ) -> BTreeMap<CodeOffset, BTreeSet<TempIndex>> {
+        let mut result = BTreeMap::new();
+        for (block_id, block_state) in state_map {
+            let mut state = block_state.pre;
+            for offset in cfg.instr_indexes(block_id).rev() {
+                let instr = &instrs[offset as usize];
+                state = self.execute(state, instr, offset);
+                result.insert(offset, state.livevars.clone());
+            }
+        }
+        result
+    }
+
+    fn execute(&mut self, pre: LiveVarState, instr: &Bytecode, _idx: CodeOffset) -> LiveVarState {
+        use Bytecode::*;
+        let mut post = pre;
+        match instr {
+            Assign(_, dst, src, _) => {
```

Comment:
Wondering whether we need to distinguish the Assign kind here (Move/Copy/Store -- where Store I think is similar to Move, as it results from popping from the stack and store in a local.)

Reply:
When writing down analyses for stackless bytecode, I try to avoid thinking about where the stackless bytecode came from. In other words, I am trying to do something sound even if the stackless bytecode was written from hand.  I want to see how far we can get with this approach before it becomes useful, if at all, to add constraints that characterize the source of the stackless bytecode.